
<!-- saved from url=(0084)https://docs.google.com/document/pub?id=1lx8cGQlyBZV8RXm7gSLqFafcOUfPq6_GjadqBxqiAzQ -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><script type="text/javascript">
(function(){(function(){function d(a){this.t={};this.tick=function(a,b,c){this.t[a]=[void 0!=c?c:(new Date).getTime(),b]};this.tick("start",null,a)}var a=new d;window.jstiming={Timer:d,load:a};if(window.performance&&window.performance.timing){var b=window.performance.timing,a=window.jstiming.load,c=b.navigationStart,b=b.responseStart;0<c&&b>=c&&(a.tick("_wtsrt",void 0,c),a.tick("wtsrt_","_wtsrt",b))}try{a=null,window.chrome&&window.chrome.csi&&(a=Math.floor(window.chrome.csi().pageT)),null==a&&window.gtbExternal&&
(a=window.gtbExternal.pageT()),null==a&&window.external&&(a=window.external.pageT),a&&(window.jstiming.pt=a)}catch(e){}})();})()
</script><script type="text/javascript">var KX_timer = new window.jstiming.Timer(); KX_timer.name = 'published';</script><title>DAY1_exercises.docx</title><style type="text/css">
      body {
        font-family: arial, sans, sans-serif;
        margin: 0;
      }

      iframe {
        border: 0;
        frameborder: 0;
        height: 100%;
        width: 100%;
      }

      #header, #footer {
        background: #f0f0f0;
        padding: 10px 10px;
      }

      #header {
        border-bottom: 1px #ccc solid;
      }

      #footer {
        border-top: 1px #ccc solid;
        border-bottom: 1px #ccc solid;
        font-size: 13;
      }

      #contents {
        margin: 6px;
      }

      .dash {
        padding: 0 6px;
      }
    </style></head><body><div id="header">DAY1_exercises.docx</div><div id="contents"><style type="text/css">@import url('https://themes.googleusercontent.com/fonts/css?kit=MSSLfUayeNh9PW3ng9UWrqo0P1CSBNc3gBWclSzSx0c');ol{margin:0;padding:0}.c24{color:#4f81bd;font-size:16pt;font-family:"Calibri"}.c26{max-width:504pt;background-color:#ffffff;padding:72pt 54pt 72pt 54pt}.c9{line-height:1.15;padding-top:0pt;padding-bottom:0pt}.c4{list-style-type:disc;margin:0;padding:0}.c14{color:inherit;text-decoration:inherit}.c7{text-indent:-49.5pt;margin-left:81pt}.c5{font-size:10pt;font-family:"Courier New"}.c3{font-size:11pt;font-family:"Courier New"}.c19{color:#0000ff;text-decoration:underline}.c25{font-size:20pt;font-family:"Calibri"}.c16{color:#1155cc;text-decoration:underline}.c10{padding-left:0pt;padding-top:0pt}.c2{padding-top:2pt}.c22{padding-left:0pt}.c12{text-align:center}.c8{height:12pt}.c21{text-indent:36pt}.c1{margin-left:36pt}.c13{font-style:italic}.c15{height:13pt}.c18{padding-top:0pt}.c17{font-weight:bold}.c20{height:16pt}.c6{text-align:left}.c23{height:11pt}.c0{direction:ltr}.c11{margin-left:28.4pt}.title{padding-top:12pt;line-height:1.0;text-align:center;color:#000000;font-size:16pt;font-family:"Arial";font-weight:bold;padding-bottom:3pt}.subtitle{padding-top:6pt;line-height:1.0;text-align:center;color:#000000;font-size:12pt;font-family:"Arial";padding-bottom:3pt}li{color:#000000;font-size:12pt;font-family:"Cambria"}p{color:#000000;font-size:12pt;margin:0;font-family:"Cambria"}h1{padding-top:24pt;line-height:1.0;text-align:justify;color:#345a8a;font-size:18pt;font-family:"Calibri";font-weight:bold;padding-bottom:0pt}h2{padding-top:18pt;line-height:1.0;text-align:justify;color:#4f81bd;font-size:16pt;font-family:"Calibri";font-weight:bold;padding-bottom:0pt}h3{padding-top:12pt;line-height:1.0;text-align:justify;color:#000000;font-size:13pt;font-family:"Arial";font-weight:bold;padding-bottom:3pt}h4{padding-top:12pt;line-height:1.0;text-align:justify;color:#000000;font-size:14pt;font-family:"Cambria";font-weight:bold;padding-bottom:3pt}h5{padding-top:12pt;line-height:1.0;text-align:justify;color:#000000;font-style:italic;font-size:13pt;font-family:"Cambria";font-weight:bold;padding-bottom:3pt}h6{padding-top:2pt;line-height:1.0;text-align:left;color:#000000;font-size:11pt;font-family:"Courier New";padding-bottom:2pt}</style><p class="c0"><span class="c17 c25">DAY 1 EXERCISES</span></p><h5 class="c0"><a name="h.qbxkf5vy9wxi"></a><span>Oksana Riba-Grognuz</span></h5><hr><h1 class="c0"><a name="h.1vvoitpdmymq"></a><span>UNIX</span></h1><p class="c0"><span>First of all let's login to Vital-IT infrastructure. Each of you received a user name to use in the secure connection command below.</span></p><p class="c0"><span class="c3">$ ssh </span><span class="c3 c17">username</span><span class="c3">@prd.vital-it.</span><span class="c3">ch</span></p><p class="c2 c8 c0"><span class="c3"></span></p><p class="c0 c2"><span>Now we are connected to a front-end node (prd.vital-it.ch) that can only be used to submit jobs to t</span><span>he </span><span>Vital-IT cluster. For this practical we will</span><span>&nbsp;</span><span>be </span><span>u</span><span>sing 2 big-memory machines of the UNIL Department of Ecology and Evolution (dee-serv01 and dee-serv02). Half of you will connect to one of these machines and half to the other. Use the machine name you were assigned below.</span></p><p class="c0"><span class="c3">$ ssh dee-serv0</span><span class="c3 c17">X</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Once you g</span><span>e</span><span>t there check where you are with the command printing the current directory.</span></p><p class="c0"><span class="c3">$ pwd</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Alternatively, use the "echo" command to find out the address of your home directory. If you did not change your directory upon login this address should be the same as when you did "pwd".</span></p><p class="c0"><span class="c3">$ echo $HOME</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>If you forgot which username you used, you can always check with the command below.</span></p><p class="c0"><span class="c3">$ whoami</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>List file and directories in your location.</span></p><p class="c0"><span class="c3">$ ls</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Wherever you are located there is always an easy way to get back to home directory. Just type "cd" (change directory) without arguments.</span></p><p class="c0"><span>How to get home?</span></p><p class="c0"><span class="c3">$ </span><span class="c3">cd</span></p><h2 class="c0"><a name="h.oga1l4b4j4xk"></a><span>Adding bioinformatics software to your $PATH</span></h2><p class="c0"><span>To see what you programs you can run simply type </span><span>the</span><span>&nbsp;TAB </span><span>key</span><span>&nbsp;twice. </span></p><p class="c0"><span class="c3">$ TAB TAB</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Possible options are the program files found in locations specified in $PATH. What are these locations? Look at $PATH with "echo" command:</span></p><p class="c0"><span class="c3">$ echo $PATH</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Right now your $PATH is missing the software we will need. &nbsp;For this one would typically do the following:</span><span>&nbsp;</span></p><p class="c0"><span class="c3">$ export PATH="$PATH:/some/additiona</span><span class="c3">l/software/</span><span class="c3">location"</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Vital-IT have prepared something that does this for you. So that it happens automatically when you login, you’ll need to edit the</span><span>&nbsp;.bashrc in </span><span>your</span><span>&nbsp;home directory.</span></p><p class="c0"><span>To see </span><span>your</span><span>&nbsp;.bashrc file with the command "ls" we need to add additional arguments, as this file is hidden by default.</span></p><p class="c0"><span class="c3">$ ls -lah</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Browse the contents of .bashrc using command "less".</span></p><p class="c0"><span class="c3">$ less .bashrc</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>D</span><span>oes</span><span>&nbsp;your .bashrc contain the following lines?</span></p><p class="c0"><span class="c3">source /mnt/common/R-BioC/R-BioC.bashrc</span></p><p class="c0"><span class="c3">source /mnt/common/DevTools/DevTools.bashrc</span></p><p class="c0"><span class="c3">source /mnt/common/UHTS/UHTS.bashrc</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c8 c0"><span></span></p><p class="c0"><span>These lines add to your $PATH the locations and configurations of R, and of &nbsp;Ultra High Throughput Sequencing (UHTS) applications. If they are missing from your .bashrc </span><span>append them to .bashrc using "&gt;&gt;"</span><span>:</span><span>&nbsp;</span></p><p class="c0"><span class="c3">$ echo</span><span class="c3">&nbsp;"</span><span class="c3">source /mnt/common/R-BioC/R-BioC.bashrc"</span><span class="c3">&nbsp;</span><span class="c3">&gt;&gt; .bashrc</span></p><p class="c0"><span class="c3">$ echo “source /mnt/common/DevTools/DevTools.bashrc” &gt;&gt; .bashrc</span></p><p class="c0"><span class="c3">$ echo</span><span class="c3">&nbsp;"</span><span class="c3">source /mnt/common/UHTS/UHTS.bashrc"</span><span class="c3">&nbsp;</span><span class="c3">&nbsp; &nbsp; &gt;&gt; .</span><span class="c3">bashrc</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c2 c8 c6 c0"><span class="c3"></span></p><p class="c2 c6 c0"><span>Now please logout and then ssh to the server again. Check how your $PATH has changed</span><span>:</span></p><p class="c0"><span class="c3">$ echo $PATH</span></p><h2 class="c0"><a name="h.b2zw2gm2xd26"></a><span>Files and folders</span></h2><p class="c0"><span>If you are not familiar with file and folder operations try the following commands.</span></p><p class="c0"><span>Ma</span><span>k</span><span>e a new empty folder.</span></p><p class="c0"><span class="c3">$ mkdir newfolder</span></p><p class="c0 c8"><span></span></p><p class="c0"><span>Go to this folder.</span></p><p class="c0"><span class="c3">$ cd newfolder</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Go one folder back.</span></p><p class="c0"><span class="c3">$ cd ..</span></p><p class="c0"><span class="c3">$ rmdir newfolder</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Make the folder again</span></p><p class="c0"><span class="c3">$ mkdir newfolder</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Create a file in that folder by redirecting printed output of "echo" command to a file.</span></p><p class="c2 c6 c0"><span class="c3">$ echo "Some text" &gt; newfolder/file.txt</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Try removing the folder again</span></p><p class="c2 c6 c0"><span class="c3">$ rmdir newfolder</span></p><p class="c8 c0"><span></span></p><p class="c1 c0"><span class="c13">Q: </span><span class="c13">Why this does not work?</span></p><p class="c8 c0"><span class="c13"></span></p><p class="c0"><span>Try command "rm" (removes files and folders)</span></p><p class="c2 c6 c0"><span class="c3">$ rm newfolder</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>This still does not work. Check in the manual (type "man" before the command) to see which parameters to put.</span></p><p class="c2 c6 c0"><span class="c3">$ man rm </span></p><p class="c2 c8 c6 c0"><span class="c17 c24"></span></p><h2 class="c2 c6 c0"><a name="h.nk5lj5r9qfpr"></a><span>Editing files</span></h2><p class="c0"><span>Files can be edited locally on Vital-IT (using nano or vi or emacs), or on your laptop using a text editor of your choice (Aquamacs, TextWrangler... NOT Microsoft Word!). To edit a file locally you must first download it from Vital-IT. For this you can use scp (or </span><span>Cyberduck or something like FileZilla).</span><span>&nbsp;</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>For example, if you choose scp do the following locally on your computer:</span></p><p class="c2 c6 c0"><span class="c3">$ mkdir Scripts</span></p><p class="c2 c6 c0"><span class="c3">$ scp </span><span class="c3 c16 c17"><a class="c14" href="mailto:username@prd.vital-it.ch">username</a></span><span class="c3 c16"><a class="c14" href="mailto:username@prd.vital-it.ch">@</a></span><span class="c3 c16"><a class="c14" href="mailto:username@prd.vital-it.ch">prd.vital-it.ch</a></span><span class="c3">:/location/of/Scripts/MyScript.rb Scripts/</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Once the script was mo</span><span>dified </span><span>upload it back to Vital-IT.</span></p><p class="c2 c6 c0"><span class="c3">$ scp Scripts/MyScript.rb </span><span class="c3 c16 c17"><a class="c14" href="mailto:username@prd.vital-it.ch">username</a></span><span class="c3 c16"><a class="c14" href="mailto:username@prd.vital-it.ch">@</a></span><span class="c3 c16"><a class="c14" href="mailto:username@prd.vital-it.ch">prd.vital-it.ch</a></span><span class="c3">:/location/of/Scripts/ </span></p><hr><h1 class="c0"><a name="h.v665mvoiy93w"></a><span>DATA ANALYSIS</span></h1><p class="c0"><span>We will be working on </span><span class="c13">Solenopsis invicta</span><span>&nbsp;(the red fire ant) using Illumina DNA and RNA-seq reads. For the official release of </span><span class="c13">de novo </span><span>genome assembly (Wurm </span><span class="c13">et al.</span><span>&nbsp;2011, PNAS) we </span><span>combined</span><span>&nbsp;Illumina and 454 technologies in a hybrid assembly approach. For this course we will </span><span>use only</span><span>&nbsp;Illumina</span><span>&nbsp;because it has become clear that it is possible to perform </span><span class="c13">de novo </span><span>genome assembly using only </span><span>this technology. Furthermore, </span><span>we are </span><span>only </span><span>considering</span><span>&nbsp;for th</span><span>is practical </span><span>a very small subset (less than 5%) of the fire ant genome. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Please form the groups of two</span><span>&nbsp;(one more computational; one less computational)</span><span>. From now on you will use only one access to </span><span>V</span><span>ital-IT per group.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Login to Vital-IT. </span><span>The command below connects first to the prd server and then to the dee server. We are only allowed to use ssh to connect to prd, but cannot use it for calculations. </span><span>&nbsp;</span></p><p class="c2 c6 c0"><span class="c3">$ ssh -t </span><span class="c3 c17">username</span><span class="c3">@prd.vital-it.ch ssh dee-serv0</span><span class="c3 c17">X</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>According to the Vital-IT rules, no calculations can be carried out in the home directory. We will use /scratch/cluster/weekly/ for all practicals (files will be deleted after one week!). Create there a directory named according to your user-name and change from your home directory to this new directory. </span></p><p class="c2 c6 c0"><span class="c3">$ mkdir /scratch/cluster/weekly/</span><span class="c3 c17">username</span><span class="c3">&nbsp;</span></p><p class="c2 c6 c0"><span class="c3">$ cd &nbsp; &nbsp;/scratch/cluster/weekly/</span><span class="c3 c17">username</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Extract the files required for today's practical. </span></p><p class="c2 c6 c0"><span class="c3">$ unzip /scratch/cluster/monthly/oribagro/summer2012_</span><span class="c3">Oksana</span><span class="c3">.zip</span></p><p class="c2 c8 c6 c0"><span class="c3"></span></p><hr><h2 class="c0"><a name="h.riy43fbmsfm6"></a><span>Quality Control: FastQC</span></h2><p class="c0"><span>W</span><span>e will use the </span><span class="c16"><a class="c14" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></span><span>&nbsp;package installed locally on your computer. We will analyse files located in DNA-seq/Raw/ and DNA-seq/Nr/. They are big. So </span><span>p</span><span>lease take a copy of th</span><span>e</span><span>se files from </span><span>provided </span><span>USB key </span><span>or from the local web server (zipped as DNA-seq.zip)</span><span>.</span><span>&nbsp;[probably </span><span class="c16"><a class="c14" href="http://192.168.167.32/">http://192.168.167.32</a></span><span>&nbsp;]</span></p><p class="c2 c6 c0"><span class="c3">$ ls DNA-seq/Raw</span></p><p class="c2 c6 c0"><span class="c3">$ ls DNA-seq/Nr</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Our goal is to make decisions on de novo assembly strategy based on FastQC quality report. Open FastQC and open files to analyse. Process first Raw files, which are subsets of Illumina Hi-seq lanes as they came out of the sequencer. </span></p><p class="c8 c0"><span></span></p><p class="c2 c0 c11"><span class="c13">Q: What do the file names mean?</span></p><p class="c2 c0 c11"><span class="c13">Q: Do you think both lanes should be used for assembly? </span></p><p class="c2 c0 c11"><span class="c13">Q: Do we need to trim or filter reads? </span></p><p class="c2 c0 c11"><span class="c13">Q: Which information is important to take the decision about trimming/filtering?</span></p><p class="c2 c0 c11"><span class="c13">Q: How can you explain a significant drop in the quality in the beginning of the reads of 2nd pair members of lane 7?</span></p><p class="c8 c0"><span class="c13"></span></p><p class="c0"><span>Process Nr reads now. </span><span>&nbsp;Nr means “Non-redundant”: </span><span>reads were processed to remove exact duplicates</span><span>.</span></p><p class="c8 c0"><span class="c13"></span></p><p class="c2 c0 c11"><span class="c13">Q: Do you see a big change in duplications levels?</span></p><p class="c2 c0 c11"><span class="c13">Q: What can be the reason for that (consider information given during presentation)?</span></p><p class="c8 c0"><span class="c13"></span></p><hr><h2 class="c0"><a name="h.6qfpahsml02f"></a><span>Quality Control: FASTX-Toolkit</span></h2><p class="c0"><span>We will use FASTX-Toolkit to implement filtering/trimming decided in the previous step. This part takes place at Vital-IT. Process Raw or Nr (non-redundant) reads and output "clean" files to /scratch/cluster/weekly/username/DNA-seq/Clean. Toolkit allows us to do different types of filtering/trimming.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Find out how to trim reads based on coordinates:</span></p><p class="c2 c6 c0"><span class="c3">$ fastx_trimmer -h</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Find out how to trim/filter reads based on quality:</span></p><p class="c2 c6 c0"><span class="c3">$ fastq_quality_trimmer –h</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Use fastx_trimmer or fastq_quality_trimmer on each file individually (replace first base -f and/or last base -l values to desired trimming in the command below). Or alternatively use a provided launch script that will do the job on all files at once.</span></p><p class="c0"><span class="c3">$ fastx_trimmer </span><span class="c3">-f </span><span class="c3">xxxx</span><span class="c3">&nbsp;-l </span><span class="c3">yyyy</span><span class="c3">&nbsp;-</span><span class="c3">i DNA-seq/Nr/101104_s_7_1.subset.fastq -o DNA-seq/Clean/101104_s_7_1.subset.fastq</span></p><p class="c0"><span class="c3">$ fastx_trimmer -f xxxx -l yyyy -i DNA-seq/Nr/101104_s_7_2.subset.fastq -o DNA-seq/Clean/101104_s_7_2.subset.fastq</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Ideally we want to remember what exactly was done to Raw data. A good way </span><span>to achieve</span><span>&nbsp;this is </span><span>by</span><span>&nbsp;doing all processing using bash wrapper scripts. You have an example of such script in Scripts folder. Another advantage of us</span><span>ing scripts that execute the command on each input file automatically, is to handle large data sets comprised of multiple files. </span></p><p class="c2 c6 c0"><span class="c3">$ less Scripts/run_fastx_trimmer.sh</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>This script should be launched in project directory containing DNA-seq directory in it. You can use vi or nano to modify this script on </span><span>Vital-IT</span><span>, or if you do not know how to use these, you can download this file to your local computer and modify using your preferred text editor to do what you judge necessary as trimming/filtering. </span><span>Hint: you need to modify the line that launches fastx_trimmer command by adjusting trimming coordinates (-f xxx and -l yyy specifying the first and the last base respectively). &nbsp;</span><span>If you forgot how to edit files, check the end of the UNIX introduction.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>To launch the script you should go to the following folder on Vital-IT:</span></p><p class="c2 c0 c6"><span class="c3">$ cd /scratch/cluster/weekly/username</span></p><p class="c2 c6 c0"><span class="c3">$ Scripts/run_fastx_trimmer.sh</span></p><h2 class="c0 c20"><a name="h.s360f8pndvrm"></a></h2><hr><h2 class="c0"><a name="h.56bza0xsuab5"></a><span>DE NOVO GENOME ASSEMBLY</span></h2><p class="c0"><span>Few definitions are important for </span><span class="c13">de novo</span><span>&nbsp;assembly: contigs (contiguous sequences) and scaffolds, illustrated in the figure below. A genome assembly consists in hundreds to thousands of scaffolds.</span></p><p class="c9 c0 c12"><img height="202" src="./DAY1_exercises.docx_files/pubimage" width="483"></p><p class="c8 c0"><span></span></p><p class="c0"><span>We will use SOAPdenovo for genome assembly. Depending on genome characteristics different software might be the most appropriate. For the red fire ant data in 2009 SOAPdenovo was the </span><span>best</span><span>&nbsp;performing assembler of Illumina reads. To keep track of our actions and ensure the reproducibility of all steps we will rely on bash scripts to run assemblies. A usual approach to Illumina assembly is to do multiple assemblies using different combinations of data quality trimming/filtering and different assembler parameters. Due to time and resource constraints, each pair of students will only pe</span><span>rform </span><span>2 or 3 assemblies. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Illumina assemblers rely on de Bruijn graph constructed from K-mers (K-length words) of all reads. This makes K-mer length a key software parameter to optimise. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>SOAPdenovo package consists of </span><span>four</span><span>&nbsp;programs: </span><span>pregraph</span><span>, contig, map and scaff. Assembly with paired end reads </span><span>involves</span><span>&nbsp;the use of all </span><span>four</span><span>&nbsp;programs</span><span>. Command "all" allows executing full SOAPdenovo package easily. We basically need to specify "SOAPdenovo all -s config_file -o output_prefix". </span></p><p class="c2 c6 c0"><span class="c3">$ cd /scratch/cluster/weekly/</span><span class="c3 c17">username</span><span class="c3">/S</span><span class="c3">OAPdenovo/Assembly</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>This folder contains </span><span>two</span><span>&nbsp;files </span><span>“</span><span>conf01-lanes47_maplen60</span><span>”</span><span>&nbsp;and </span><span>“</span><span>Run_conf01-RL200D.sh</span><span>”</span><span>.</span></p><p class="c2 c6 c0"><span class="c3">$ less </span><span class="c3">c</span><span class="c3">onf01-lanes47_maplen60</span></p><p class="c2 c6 c0"><span class="c3">$ less Run_conf01-RL200D.sh</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Don’t run anything yet! </span><span>Config file is created according to a format defined in SOAPdenovo requirements. Run_conf01-RL200D.sh will use config file to run SOAPdenovo with the specified parameters and config files and output results to folders named according to parameter values.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Check the explanations of commands and config file specifications at </span><span class="c19"><a class="c14" href="http://soap.genomics.org.cn/soapdenovo.html">http://soap.genomics.org.cn/soapdenovo.html</a></span><span>&nbsp;.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Modify </span><span class="c3">conf01-lanes47_maplen60. </span><span>You need at least to specify the correct locatio</span><span>n of input fastq files that you want to use </span><span>(replace username). Optionally you can change some of the parameters.</span></p><p class="c0"><span>&nbsp;</span></p><p class="c2 c0 c11"><span class="c13">Q: Can you trim reads within SOAPdenovo config file?</span><span class="c13">&nbsp;</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Modify </span><span>Run_conf01-RL200D</span><span>.sh. Use K-mer values of at least 1/3 of read length. Do not put more than </span><span>three</span><span>&nbsp;K-mer values as it will increase run time (</span><span>Please do not run more than one assembly at a time - </span><span>&nbsp;we are a big group of students sharing </span><span>few compute resources).</span><span>&nbsp;</span></p><p class="c8 c0"><span></span></p><p class="c2 c0 c11"><span class="c13">Q: What are the different parameters used to run SOAPdenovo in Run_conf01-RL200D.sh</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>After you have edited the script, make sure you are in the correct folder and launch it:</span></p><p class="c2 c6 c0"><span class="c3">$ ./Run_conf01-RL200D.sh</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>After the assembly is finished examine the contents of output folders.</span><span>&nbsp;</span><span>Look at the contents of LOG file form the assembly</span></p><p class="c2 c6 c0"><span class="c3">$ less LOG</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Files out.contig and out.scafSeq respectively contain F</span><span>ASTA format </span><span>scaffold an</span><span>d</span><span>&nbsp;contig sequences.</span></p><h3 class="c0"><a name="h.15oti3d6fvbz"></a><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Optional</span></h3><p class="c1 c0"><span class="c13">Q: What does asm_flags=3 in the config file mean?</span></p><p class="c1 c0"><span>There are multiple versions of SOAPdenovo. To see them, do:</span></p><p class="c1 c0"><span class="c3">$ SOAPdenovo TAB TAB</span></p><p class="c1 c0"><span class="c13">Q: Why are there several versions? When is it dangerous to use </span><span class="c13">SOAPdenovo-127mer</span><span class="c13">?</span></p><h2 class="c20 c0"><a name="h.hmj0ww8r2nsc"></a></h2><hr><h2 class="c20 c0"><a name="h.tpt1n8gxlh61"></a></h2><h1 class="c0"><a name="h.lv3bs5x7vqq5"></a><span>ASSESSING ASSEMBLY QUALITY</span></h1><h2 class="c0"><a name="h.qexprhun430c"></a><span>Comparing</span><span>&nbsp;</span><span>assembly</span><span>&nbsp;metrics</span></h2><p class="c0"><span>A common way to select the optimal assembly strategy is to look at various types of statistics, like total number of assembled base pairs, number of scaffolds, N50 </span><span>length</span><span>, maximum scaffold length etc.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>A script distributed with the assembly software Abyss called "abyss-fac" allows generating this kind of statistics. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Run abyss-fac script on results on one of your scaffolds files (replace the last assembly folder with the one you generated).</span></p><p class="c2 c6 c0"><span class="c3">$ cd /scratch/cluster/weekly/username/SOAPdenovo/Assembly/conf01_K35_R_L200_D</span></p><p class="c2 c6 c0"><span class="c3">$ abyss-fac out.scafSeq</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Now run program called seqstat to generate statistics.</span></p><p class="c2 c6 c0"><span class="c3">$ seqstat out.scafSeq</span></p><p class="c2 c8 c0 c11"><span class="c3"></span></p><p class="c2 c0 c11"><span class="c13">Q: Why do some numbers of seqstat differ from those generated with abyss-fac?</span></p><p class="c2 c0 c11"><span class="c13">Q: Which do you find the most appropriate to use in selecting best assembly?</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>We can compare statistics between contig fasta and scaffold fasta using a cutoff equal to the value of -L parameter used in the assembly.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Run abyss-fac script on scaffolds again specifying -t 200 (the value of -L used in the assembly) as a cutoff for minimum scaffold length.</span></p><p class="c2 c6 c0"><span class="c3">$ abyss-fac -t 200 out.scafSeq</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Do the same for contigs</span></p><p class="c2 c6 c0"><span class="c3">$ abyss-fac -t 200 out.contig</span></p><p class="c2 c8 c6 c0"><span class="c3"></span></p><p class="c2 c1 c6 c0"><span class="c13">Q: Why is the number of bp not the same? </span></p><p class="c2 c8 c1 c6 c0"><span></span></p><h3 class="c1 c0"><a name="h.9nnv646vjl16"></a><span>Optional (crucial for submitting genome assembly to NCBI) </span></h3><p class="c1 c0"><span>In order to submit a newly assembled genome to a public database one needs to submit contig sequences and an AGP file specifying how these sequences are arranged in scaffolds. Scaffolds are built from contigs (contiguous sequences), that are joined by stretches of N bases using the information form paired reads with known insert size. Unfortunately SOAPdenovo, as most other de Brujin graph based programs does not provide such output and report a contig file with all sequences (regardless of inclusion in scaffolds). </span><span>We can generate an AGP file containing only contigs that were used for scaffold construction by defining contigs based on stretches of N sequences within scaffolds.</span></p><p class="c2 c1 c6 c0"><span class="c3">$ /scratch/cluster/weekly/username/Scripts/fasta2agp.pl -f out.scafSeq -p out</span></p><p class="c8 c1 c0"><span></span></p><p class="c1 c0"><span>This will generate files out.contigs.fa and out.agp. Run abyss-fac on new contigs file</span></p><p class="c2 c1 c6 c0"><span class="c3">$</span><span class="c3">&nbsp;abyss-fac -t 200 out.contigs.fa</span></p><p class="c8 c1 c0"><span></span></p><p class="c1 c0"><span>Now the total number of bp is the same as in scaffold fasta file</span><span>.</span><span>&nbsp;</span></p><p class="c1 c0"><span>Have a look at AGP file. Read about format specifications at the following link: </span><span class="c19"><a class="c14" href="http://www.ncbi.nlm.nih.gov/projects/genome/assembly/agp/AGP_Specification.shtml">http://www.ncbi.nlm.nih.gov/projects/genome/assembly/agp/AGP_Specification.shtml</a></span><span>&nbsp; </span></p><p class="c2 c1 c6 c0"><span class="c3">$ less out.agp</span></p><p class="c8 c0"><span class="c3"></span></p><p class="c0"><span>Often one would generate multiple assemblies to refine the strategy (</span><span>many</span><span>&nbsp;more than </span><span>three</span><span>). Although we have generated 2-3 assemblies, we will act as if we had a lot and process statistics from these files into R-readable format. We will save the file in folder called SOAPdenovo/Statistics .</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">cd /scratch/cluster/weekly/username/SOAPdenovo/Assembly</span></p><p class="c2 c6 c0"><span class="c3">$ ./get.stats.sh &gt;</span><span class="c3">&nbsp;</span><span class="c3">stats</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Now</span><span>&nbsp;use scp or other means to get file </span><span>“stats” </span><span>locally on your computer to use with R. If you </span><span>obtained no statistics, modify the read.table command in what follows by replacing “stats” with &nbsp;“</span><span class="c5 c16"><a class="c14" href="http://www.antgenomes.org/~yannickwurm/tmp/manyAssemblyStats.txt">http://www.antgenomes.org/~yannickwurm/tmp/manyAssemblyStats.txt</a></span><span class="c5">”. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>&nbsp;Lets use R to find the best assembly in terms of quantitative metrics. Locally on your computer:</span></p><h6 class="c0 c23"><a name="h.ir57ecmnmvsd"></a></h6><p class="c2 c6 c0"><span class="c3">$ R</span></p><p class="c7 c2 c6 c0"><span class="c5">s</span><span class="c5">tats</span><span class="c5">&nbsp; &nbsp; &lt;- read.table("stats", header=T, sep="\t")</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">## get K-mer values from file name</span></p><p class="c7 c2 c6 c0"><span class="c5">myKmers &nbsp; &nbsp;&lt;- substr(stats$file, 8,10)</span></p><p class="c7 c2 c6 c0"><span class="c5">## get config names</span></p><p class="c7 c2 c6 c0"><span class="c5">myConf &nbsp; &nbsp; &lt;- as.numeric(substr(stats$file, 5,6))</span></p><p class="c7 c2 c6 c0"><span class="c5">## get all color names containing string "dark"</span></p><p class="c7 c2 c6 c0"><span class="c5">allColors &nbsp;&lt;- colors()[grep("dark",colors())]</span></p><p class="c7 c2 c6 c0"><span class="c5">## randomly sample the neccessary number of colors</span></p><p class="c7 c2 c6 c0"><span class="c5">myColors &nbsp; &lt;- sample(allColors, length(unique(myConf)))</span></p><p class="c7 c2 c6 c0"><span class="c5">## make vector of colors per assembly setting</span></p><p class="c7 c2 c6 c0"><span class="c5">confColors &lt;- myColors[as.factor(myConf)]</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">configName &lt;- paste("config", as.character(unique(myConf)), sep="")</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">## Lets plot this:</span></p><p class="c7 c2 c6 c0"><span class="c5">myTitle &lt;- "N50 vs Number of contigs &gt;= 200bp\nSize is proportional to total assembly bp"</span></p><p class="c7 c2 c6 c0"><span class="c5">plot(x &nbsp; &nbsp;= stats$n.200,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;y &nbsp; &nbsp;= stats$N50,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;cex &nbsp;= (stats$sum/1000000), &nbsp;# circle size</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;col &nbsp;= confColors,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;pch &nbsp;= 19, &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; # symbol "filled circle"</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;main = myTitle,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;xlab = "Number of Contigs &gt; = 200 bp",</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;ylab = "N50")</span></p><p class="c7 c2 c6 c0"><span class="c5">legend("topright", configName, col=unique(confColors), pch=19)</span></p><p class="c7 c2 c6 c0"><span class="c5">text(stats$n.200, stats$N50, myKmers, cex=0.7, col="white")</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">myTitle &lt;- "N50 vs Number of contigs &gt;= N50\nSize is proportional to total assembly bp"</span></p><p class="c7 c2 c6 c0"><span class="c5">plot(x &nbsp; &nbsp;= stats$n.N50,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;y &nbsp; &nbsp;= stats$N50,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;cex &nbsp;= (stats$sum/1000000), </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;col &nbsp;= confColors,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;pch &nbsp;= 19,</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;main = myTitle, </span></p><p class="c2 c6 c0 c7"><span class="c5">&nbsp; &nbsp; &nbsp;xlab = "Number of Contigs &gt;= N50",</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;ylab ="N50")</span></p><p class="c7 c2 c6 c0"><span class="c5">legend("topright", configName, col=unique(confColors), pch=19)</span></p><p class="c7 c2 c6 c0"><span class="c5">text(stats$n.N50, stats$N50, myKmers, cex=0.7, col="white")</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">## Finally make a plot with the values we expect to get. </span></p><p class="c7 c2 c6 c0"><span class="c5">## The same subset of data represented 3066758 bp accross eight</span></p><p class="c7 c2 c6 c0"><span class="c5">## scaffolds with N50 of </span><span class="c5">1867187</span></p><p class="c7 c2 c6 c0"><span class="c5">myTitle &lt;- "N50 vs Number of contigs &gt;= 200bp\nSize is proportional to total assembly bp"</span></p><p class="c7 c2 c6 c0"><span class="c5">plot(x &nbsp; &nbsp;= c(stats$n.200, 8), </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;y &nbsp; &nbsp;= c(stats$N50, </span><span class="c5">1867187</span><span class="c5">), </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;cex &nbsp;= c(stats$sum, 3066758)/1000000, </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;col &nbsp;= c(confColors, "red"),</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;pch &nbsp;= 19, </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;main = myTitle, </span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;xlab = "Number of Contigs &gt;= 200 bp",</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp;ylab ="N50")</span></p><p class="c7 c2 c8 c6 c0"><span class="c5"></span></p><p class="c7 c2 c6 c0"><span class="c5">legend("topright",</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;c(configName, "Official"),</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;col=unique(c(confColors,"red")),</span></p><p class="c7 c2 c6 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;pch=19)</span></p><p class="c7 c2 c8 c6 c0"><span class="c3"></span></p><p class="c7 c2 c8 c6 c0"><span class="c3"></span></p><p class="c2 c8 c0 c11"><span class="c3"></span></p><p class="c2 c0 c11"><span class="c13">Q: Which assembly is the best in terms of quantitative metrics?</span></p><p class="c2 c0 c11"><span class="c13">Q: Why there is a significant difference with official release?</span></p><hr><h2 class="c0"><a name="h.mggk6cxnuiix"></a><span>Quality assessment using independent information</span></h2><p class="c0"><span>Independently obtained information can provide the most reliable measures of whether or not your assembly is accurate. </span></p><ol class="c4" start="1"><li class="c1 c0 c22"><span>One approach involves mapping to your genome </span><span class="c16"><a class="c14" href="http://korflab.ucdavis.edu/Datasets/cegma/">a conserved set of “core eukaryotic genes” </a></span><span>that exist in exactly one single full copy in a broad range of eukaryotic genomes. Deviations from this provides quantitative estimations of the accuracy and likely &nbsp;completeness of your genome. </span></li><li class="c22 c1 c0"><span>Similar approaches can be used by mapping genome or proteome data from closely related species to your genome. </span></li><li class="c22 c1 c0"><span>Below, we use </span><span class="c13">de novo </span><span>assembled </span><span>transcriptome information that we independently obtained from the same species, </span><span class="c13">Solenopsis invicta. </span><span>That will require first obtaining a transcriptome assembly.</span></li></ol><hr><h2 class="c0"><a name="h.hnwv2c7o2xeh"></a><span>DE NOVO TRANSCRIPTOME ASSEMBLY</span></h2><p class="c0"><span>We will use </span><span class="c16"><a class="c14" href="http://trinityrnaseq.sourceforge.net/">the Trinity software</a></span><span>&nbsp;to assemble transcriptome from single Illumina reads</span><span>:</span><span>&nbsp;</span></p><p class="c2 c6 c0"><span class="c3">$ cd /scratch/cluster/weekly/username/Trinity</span></p><p class="c0"><span>First, check that your .</span><span class="c3">bashrc</span><span>&nbsp;contains the following line (if not, please add it)</span></p><p class="c0 c21"><span class="c3">source /mnt/common/DevTools/DevTools.bashrc</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Have a look at options of Trinity.pl:</span></p><p class="c2 c6 c0"><span class="c3">$ Trinity.pl -h</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>We have RNA-seq data for 3 conditions: males, queens and workers prefixed M, Q and W, respectively. Each condition is present in 4 replicates, numbers in file names indicate the ant colony from which the sample was taken. Due to time constraints we will restrict </span><span class="c13">de novo</span><span>&nbsp;assembly to a single RNA-seq file. </span><span>Choose one RNA-seq file in /scratch/cluster/weekly/usernameRNA-seq/Raw/ and run Trinity on it. </span><span>&nbsp;You can s</span><span>tore the file name in variable raw</span><span>Rnase</span><span>qFile</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">rawRNAseqFile</span><span class="c3">=/scratch/cluster/weekly/username/RNA-seq/Raw/Q415.subset.fastq</span></p><p class="c2 c6 c0"><span class="c3">$ Trinity.pl --seqType fq --kmer_method jellyfish --max_memory 25G --CPU 1 --bflyCalculateCPU --single $</span><span class="c3">rawRNAseqFile</span><span class="c3">&nbsp;&nbsp;&gt; stdout 2&gt; stderr</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>If you do not manage to run Trinity, then copy its output as described below (make sure you are in the directory called “Trinity”):</span></p><p class="c6 c0"><span class="c3">$ mkdir &nbsp;-p trinity_out_dir</span></p><p class="c6 c0"><span class="c3">$ cp /scratch/cluster/monthly/oribagro/summer2012/Trinity/trinity_out_dir/Trinity.fasta trinity_out_dir/</span></p><p class="c8 c6 c0"><span class="c3"></span></p><p class="c0"><span>Find out how many transcripts were assembled using “seqstat” or “abyss-fac”.</span></p><p class="c2 c6 c0"><span class="c3">$ seqstat trinity_out_dir/Trinity.fasta</span></p><h3 class="c2 c1 c15 c6 c0"><a name="h.gfy25qroohe9"></a></h3><h3 class="c2 c1 c6 c0"><a name="h.wyfbuddpkjlc"></a><span>Optional</span></h3><p class="c1 c0"><span class="c13">Q: How many genes are represented in the fasta file? Are you sure?</span></p><p class="c0 c1"><span class="c13">Q:</span><span class="c13">&nbsp;</span><span class="c13">What are the different steps of the Trinity pipeline?</span></p><p class="c8 c0"><span></span></p><hr><h2 class="c0"><a name="h.sdncdyh6n5ts"></a><span>CROSS-VALIDATION OF DE NOVO ASSEMBLIES</span></h2><p class="c0"><span>You will align transcripts assembled with Trinity to the genomic assembly that is best according to quantitative quality metrics. We will use a script that uses the </span><span class="c16"><a class="c14" href="http://genome.ucsc.edu/FAQ/FAQblat.html">BLAT</a></span><span class="c16"><a class="c14" href="http://genome.ucsc.edu/FAQ/FAQblat.html">&nbsp;aligner</a></span><span>&nbsp;to generate alignment statistics and to construct network files. Network files include all cases where transcript is partially aligned on different scaffolds (non overlapping, complementary parts of transcript). </span><span>Subsequently</span><span>&nbsp;we </span><span>will</span><span>&nbsp;visualise the order inferred form such transcript alignments and the order of contigs in scaffolds. See more details at </span><span class="c19"><a class="c14" href="https://github.com/ksanao/TGNet">https://github.com/ksanao/TGNet</a></span><span>&nbsp;.</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Generate AGP file with scaffold structure (if you haven't already done so yet for the best assembly)</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">bestAssembly=conf01_K35_R_L200_D</span></p><p class="c2 c6 c0"><span class="c3">$ runDir=/scratch/cluster/weekly/</span><span class="c3 c17">username</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">cd</span><span class="c3">&nbsp;</span><span class="c3">"$runDir"</span><span class="c3">/SOAPdenovo/Assembly/"$bestAssembly"</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">"$runDir"/</span><span class="c3">Scripts/fasta2agp.pl -f out.scafSeq -p out</span></p><p class="c2 c6 c0"><span class="c3">$ </span><span class="c3">cd "$runDir"</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>We will use a </span><span>script that wraps the necessary commands</span><span>: scaffold and contig files will be processed individually and result in two networks</span><span>. </span><span>Have a look in the script. The second paragraph of this script specifies that it will not execute unless provided a directory of best assembly. Try launching it without arguments.</span></p><p class="c2 c6 c0"><span class="c3">$</span><span class="c3">&nbsp;less </span><span class="c3">Scripts/</span><span class="c3">run_TGNet</span><span class="c3">.sh</span></p><p class="c2 c6 c0"><span class="c3">$</span><span class="c3">&nbsp;Scripts/run_TGNet.sh</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Finally launch the script providing it the directory with best assembly.</span></p><p class="c2 c6 c0"><span class="c3">$ Scripts/run_TGNet.sh "$bestAssembly"</span></p><p class="c2 c6 c0"><span class="c3">$ cd SOAPdenovo/Validation/"$bestAssembly"</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>The </span><span class="c3">log.contigs </span><span>and </span><span class="c3">log.scaffolds </span><span>files summarize what was done. </span></p><p class="c0"><span>These files mention additional output files that will help you examine the following: </span></p><p class="c0"><span class="c3">$ less out.contig.blat.stat</span></p><p class="c2 c6 c0"><span class="c3">$ less out.scaffold.blat.stat</span></p><p class="c2 c8 c6 c0"><span class="c3 c17"></span></p><p class="c2 c0 c11"><span class="c13">Q:</span><span class="c13">&nbsp;How do you explain not all transcripts match over their full length? </span></p><p class="c2 c0 c11"><span class="c13">Q: </span><span class="c13">How do the statistics of mapping to scaffold and mapping to contigs differ? Why? What could be done to improve these numbers?</span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Choose one of the transcripts in cat out.scaffold.nw network file (column 2, tab delimited) and look up the alignments for this transcript in filtered and non-filtered blast output</span></p><p class="c2 c6 c0"><span class="c3">$ myTranscript=`tail -1 out.scaffold.nw | cut -f2`</span></p><p class="c2 c6 c0"><span class="c3">$ echo $myTranscript</span></p><p class="c2 c6 c0"><span class="c3">$ grep $myTranscript out.scaffold.psl</span></p><p class="c2 c6 c0"><span class="c3">$ grep $myTranscript out_filtered.scaffold.psl</span></p><p class="c2 c8 c0 c11"><span class="c13"></span></p><p class="c2 c0 c11"><span class="c13">Q: What does i</span><span class="c13">t mean when a transcripts aligns to multiple scaffolds or contigs? How can this information be useful?</span></p><p class="c8 c6 c0 c9"><span></span></p><p class="c9 c6 c0"><span>Now download the </span><span>whole </span><span class="c3">/scratch/cluster/weekly/username/SOAPdenovo/Validation</span><span>&nbsp;folder (including the network files you just created in </span><span class="c3">"$bestAssembly"</span><span>&nbsp;and </span><span class="c3">TGNet.props)</span><span>&nbsp;</span><span>to </span><span>your local computer for Cytoscape visualisation.</span></p><p class="c2 c8 c6 c0"><span></span></p><p class="c0"><span>Run Cytoscape and import network files and visual properties file first for contigs network, then for scaffolds as described below. </span></p><p class="c8 c0"><span></span></p><p class="c0"><span>Importing network</span></p><ol class="c4" start="1"><li class="c10 c1 c0"><span>From cytoscape menu select File, then Import and choose Network from Table option in drop-down menu</span></li></ol><ol class="c4" start="1"><li class="c10 c1 c6 c0"><span>Select file out.scaffold.nw (for scaffolds) or out.contig.nw (for contigs)</span></li><li class="c10 c1 c6 c0"><span>Specify columns 1,2,3 as target, interaction, source respectfully </span><span>(ie. backwards!)</span><span>, and </span><span>enable</span><span>&nbsp;all other columns</span></li><li class="c10 c1 c6 c0"><span>Tick check-box Show Text File Import Options</span></li><li class="c10 c1 c6 c0"><span>Tick check-box Transfer first line as attribute names</span></li><li class="c10 c1 c6 c0"><span>Import</span></li></ol><p class="c8 c0"><span></span></p><p class="c0"><span>Importing node attributes</span></p><ol class="c4" start="6"><li class="c10 c1 c6 c0"><span>From Cytoscape menu select File, then Import and choose Attribute from Table option in drop-down menu</span></li><li class="c10 c1 c6 c0"><span>Select file out.scaffold.attr (for scaffolds) or out.contig.attr (for contigs)</span></li><li class="c10 c1 c6 c0"><span>Tick check-box Show Text File Import Options</span></li><li class="c10 c1 c6 c0"><span>Tick check-box Transfer first line as attribute names</span></li><li class="c10 c1 c6 c0"><span>Import</span></li></ol><p class="c8 c0"><span></span></p><p class="c0"><span>Importing visual style</span></p><ol class="c4" start="11"><li class="c10 c1 c6 c0"><span>From Cytoscape menu select File, then Import and choose Vizmap Property File</span></li><li class="c10 c1 c6 c0"><span>Import a pre-built file TGNet.props</span></li><li class="c1 c6 c0 c10"><span>From VizMapper menu select a style called TGNet</span></li></ol><p class="c8 c6 c0 c18"><span></span></p><p class="c18 c6 c0"><span>Network L</span><span>ayout</span></p><ol class="c4" start="14"><li class="c10 c1 c6 c0"><span>From Cytoscape menu select Layout, then Cytoscape Layouts, then choose Force-Directed Layout, unweighted</span></li></ol><p class="c8 c0"><span></span></p><p class="c0"><span>Save your session (file.cys</span><span>).</span></p><p class="c2 c0 c11"><span class="c13">Q: </span><span class="c13">Can transcript alignments confirm scaffolding in contigs network?</span></p><p class="c2 c0 c11"><span class="c13">Q: Can your scaffolds be extended based on transcript alignment displayed in scaffold network?</span></p><p class="c2 c0 c11"><span class="c13">Q: Do you have any problematic regions, potentially misassembled?</span></p><h3 class="c1 c0 c15"><a name="h.bdj28rklymvw"></a></h3><h3 class="c1 c0"><a name="h.bfy2khx6vl9g"></a><span>Optional: </span></h3><p class="c1 c0"><span class="c13">Q: What else can Cytoscape be used for?</span><span>&nbsp;</span></p><p class="c8 c1 c0"><span></span></p></div><div id="footer"><span>Published by <a target="_blank" title="Learn more about Google Docs" href="https://docs.google.com/">Google Docs</a></span><span class="dash">–</span><a href="https://docs.google.com/abuse?id=1lx8cGQlyBZV8RXm7gSLqFafcOUfPq6_GjadqBxqiAzQ">Report Abuse</a><span class="dash">–</span><span>Updated automatically every 5 minutes</span></div><script type="text/javascript">
(function(){if(window.jstiming){window.jstiming.c={};window.jstiming.f=1;var e=function(b,a,d){var c=b.t[a],g=b.t.start;if(c&&(g||d))return c=b.t[a][0],void 0!=d?g=d:g=g[0],c-g},l=function(b,a,d){var c="";window.jstiming.pt&&(c+="&srt="+window.jstiming.pt,delete window.jstiming.pt);try{window.external&&window.external.tran?c+="&tran="+window.external.tran:window.gtbExternal&&window.gtbExternal.tran?c+="&tran="+window.gtbExternal.tran():window.chrome&&window.chrome.csi&&(c+="&tran="+window.chrome.csi().tran)}catch(g){}var f=
window.chrome;if(f&&(f=f.loadTimes)){f().wasFetchedViaSpdy&&(c+="&p=s");if(f().wasNpnNegotiated){var c=c+"&npn=1",i=f().npnNegotiatedProtocol;i&&(c+="&npnv="+(encodeURIComponent||escape)(i))}f().wasAlternateProtocolAvailable&&(c+="&apa=1")}var j=b.t,p=j.start,f=[],i=[],h;for(h in j)if("start"!=h&&0!=h.indexOf("_")){var k=j[h][1];k?j[k]&&i.push(h+"."+e(b,h,j[k][0])):p&&f.push(h+"."+e(b,h))}delete j.start;if(a)for(var m in a)c+="&"+m+"="+a[m];(a=d)||(a="https:"==document.location.protocol?"https://csi.gstatic.com/csi":
"http://csi.gstatic.com/csi");return[a,"?v=3","&s="+(window.jstiming.sn||"_s")+"&action=",b.name,i.length?"&it="+i.join(","):"","",c,"&rt=",f.join(",")].join("")};window.jstiming.getReportUri=l;var n=function(b,a,d){b=l(b,a,d);if(!b)return"";var a=new Image,c=window.jstiming.f++;window.jstiming.c[c]=a;a.onload=a.onerror=function(){window.jstiming&&delete window.jstiming.c[c]};a.src=b;a=null;return b};window.jstiming.report=function(b,a,d){if("prerender"==document.webkitVisibilityState){var c=!1,g=
function(){if(!c){a?a.prerender="1":a={prerender:"1"};var f;"prerender"==document.webkitVisibilityState?f=!1:(n(b,a,d),f=!0);f&&(c=!0,document.removeEventListener("webkitvisibilitychange",g,!1))}};document.addEventListener("webkitvisibilitychange",g,!1);return""}return n(b,a,d)};var o=function(b,a,d,c){return 0<d?(c?b.tick(a,c,d):b.tick(a,"",d),!0):!1};window.jstiming.getNavTiming=function(b){if(window.a&&window.a.e){var a=window.a.e;o(b,"_dns",a.j)&&o(b,"dns_",a.i,"_dns");o(b,"_con",a.h)&&o(b,"con_",
a.g,"_con");o(b,"_req",a.n)&&o(b,"req_",a.b,"_req");o(b,"_rcv",a.b)&&o(b,"rcv_",a.o,"_rcv");if(o(b,"_ns",a.d)){o(b,"ntsrt_",a.b,"_ns");o(b,"nsfs_",a.k,"_ns");var d=window.external&&window.external.startE;!d&&(window.chrome&&window.chrome.csi)&&(d=Math.floor(window.chrome.csi().startE));d&&(o(b,"_se",d),o(b,"sens_",a.d,"_se"));o(b,"ntplt0_",a.m,"_ns");o(b,"ntplt1_",a.l,"_ns")}}}};})()
</script><script type="text/javascript">KX_timer.tick('tl'); if (document.location.protocol == 'https:') {window.jstiming.report(KX_timer, undefined , 'https://gg.google.com/csi');} else {window.jstiming.report(KX_timer);}</script></body><link rel="stylesheet" type="text/css" href="data:text/css,"></html>